src/python/pants/source/source_root.py

```diff
@@ -29,15 +31,20 @@ class AllSourceRoots(Collection[SourceRoot]):
 class SourceRoots:
     """An interface for querying source roots."""
 
-    def __init__(self, source_root_config: "SourceRootConfig") -> None:
+    def __init__(
+        self,
+        roots: List[str],
+        fail_if_unmatched: bool = True,
+        source_root_config: Optional["SourceRootConfig"] = None,
+    ) -> None:
         """Create an object for querying source roots via patterns in a trie.
 
-        :param source_root_config: The SourceRootConfig for the source root patterns to query against.
-
         Non-test code should not instantiate directly. See SourceRootConfig.get_source_roots().
         """
-        self._trie = source_root_config.create_trie()
-        self._options = source_root_config.get_options()
+        self._roots: Set[PurePath] = set(PurePath(root) for root in roots)
```

Comment:
Is MyPy making you use this type hint? It should be inferrable.

Also, consider using `FrozenOrderedSet`. While I don't think this uses the engine, it's a good default and it behaves very very similarly to a `set`, like being able to do `fos1 | fos2` still.

Reply:
We cannot right now because `add_source_root()` mutates self._roots. And even if we could, seems like overkill to the casual reader.